Mathlib/CategoryTheory/Sites/LocallySurjective.lean

```diff
@@ -153,26 +257,97 @@ noncomputable def sheafificationIsoImagePresheaf :
+    IsLocallySurjective J (J.toSheafify P) := by
+  dsimp [GrothendieckTopology.toSheafify]
+  rw [GrothendieckTopology.plusMap_toPlus]
+  infer_instance
+
+instance isLocallySurjective_toSheafify' {D : Type*} [Category D]
+    [ConcreteCategory.{max u v} D]
+    (P : Cᵒᵖ ⥤ D) [HasWeakSheafify J D] [J.HasSheafCompose (forget D)]
+    [J.PreservesSheafification (forget D)] :
+    IsLocallySurjective J (toSheafify J P) := by
+  rw [isLocallySurjective_iff_whisker_forget,
+    ← sheafComposeIso_hom_fac, ← toSheafify_plusPlusIsoSheafify_hom]
+  infer_instance
+#align category_theory.to_sheafify_is_locally_surjective CategoryTheory.Presheaf.isLocallySurjective_toSheafify'
 
 end
 
+end Presheaf
+
+namespace Sheaf
+
+variable {J}
+variable {F₁ F₂ F₃ : Sheaf J A} (φ : F₁ ⟶ F₂) (ψ: F₂ ⟶ F₃)
+
+/-- If `φ : F₁ ⟶ F₂` is a morphism of sheaves, this is an abbreviation for
+`Presheaf.IsLocallySurjective J φ.val`. -/
+abbrev IsLocallySurjective := Presheaf.IsLocallySurjective J φ.val
+
+instance isLocallySurjective_comp [IsLocallySurjective φ] [IsLocallySurjective ψ] :
+    IsLocallySurjective (φ ≫ ψ) :=
+  Presheaf.isLocallySurjective_comp J φ.val ψ.val
+
+instance isLocallySurjective_of_iso [IsIso φ] : IsLocallySurjective φ := by
+  have : IsIso φ.val := (inferInstance : IsIso ((sheafToPresheaf J A).map φ))
+  infer_instance
+
+instance {F G : Sheaf J (Type w)} (f : F ⟶ G) :
+    IsLocallySurjective (toImageSheaf f) := by
+  dsimp [toImageSheaf]
+  infer_instance
+
+variable [J.HasSheafCompose (forget A)]
+
+instance [IsLocallySurjective φ] :
+    IsLocallySurjective ((sheafCompose J (forget A)).map φ) :=
+  (Presheaf.isLocallySurjective_iff_whisker_forget J φ.val).1 inferInstance
+
+theorem isLocallySurjective_iff_isIso {F G : Sheaf J (Type w)} (f : F ⟶ G) :
+    IsLocallySurjective f ↔ IsIso (imageSheafι f) := by
+  dsimp only [IsLocallySurjective]
+  rw [imageSheafι, Presheaf.isLocallySurjective_iff_imagePresheaf_sheafify_eq_top',
+    Subpresheaf.eq_top_iff_isIso]
+  exact isIso_iff_of_reflects_iso (f := imageSheafι f) (F := sheafToPresheaf J (Type w))
+#align category_theory.is_locally_surjective_iff_is_iso CategoryTheory.Sheaf.isLocallySurjective_iff_isIso
+
+instance epi_of_isLocallySurjective' {F₁ F₂ : Sheaf J (Type w)} (φ : F₁ ⟶ F₂)
+    [IsLocallySurjective φ] : Epi φ where
+  left_cancellation {Z} f₁ f₂ h := by
+    ext X x
+    apply (Presieve.isSeparated_of_isSheaf J Z.1 ((isSheaf_iff_isSheaf_of_type _ _).1 Z.2) _
+      (Presheaf.imageSieve_mem J φ.val x)).ext
+    rintro Y f ⟨s : F₁.val.obj (op Y), hs : φ.val.app _ s = F₂.val.map f.op x⟩
+    dsimp
+    have h₁ := congr_fun (f₁.val.naturality f.op) x
+    have h₂ := congr_fun (f₂.val.naturality f.op) x
+    dsimp at h₁ h₂
+    rw [← h₁, ← h₂, ← hs]
+    exact congr_fun (congr_app ((sheafToPresheaf J _).congr_map h) (op Y)) s
+
+instance epi_of_isLocallySurjective [IsLocallySurjective φ] : Epi φ :=
+  (sheafCompose J (forget A)).epi_of_epi_map inferInstance
+
```

Comment:
Can you also add the inverse direction?

Reply:
(epi -> locally surjective)

Reply:
Yes, now that #13320 was merged, I should be able to add this for sheaves of types. (For a few more general categories, this may need #13324.)

Reply:
I have just added `isLocallySurjective_iff_epi`.

Reply:
Yes, I had a very draft PR proving this https://github.com/leanprover-community/mathlib4/pull/9965/files#diff-cc5a48855a86465d41fe5fb49fedf7ba290877da802940a3b69530d5660ea606R150-R152 until I found you had already done it (in a much better way than I would have...).